text/0000-macros-1.1.md

```diff
@@ -0,0 +1,552 @@
+number of functions which act as token stream transformers to implement macro
+functionality.
+
+A macro crate might look like:
+
+```rust
+#![crate_type = "rustc-macro"]
+#![crate_name = "double"]
+
+extern crate rustc_macro;
+
+use rustc_macro::{Context, TokenStream};
+
+#[rustc_macro_derive(Double)]
+pub fn double(cx: &mut Context, input: TokenStream) -> TokenStream {
+    let source = input.to_source(cx);
+
+    // Parse `source` for struct/enum declaration, and then build up some new
+    // source code representing representing a number of items in the
+    // implementation of the `Double` trait for the struct/enum in question.
+    let source = derive_double(cx, source);
+
+    // Parse this back to a token stream and return it
+    TokenStream::from_source(cx, &source).unwrap()
+}
+```
+
+This new `rustc_macro_derive` attribute will be allowed inside of a
+`rustc-macro` crate but disallowed in other crate types. It defines a new
+`#[derive]` mode which can be used in a crate. The input here is the entire
+struct that `#[derive]` was attached to, attributes and all. The output is
+**expected to include the `struct`/`enum` itself** as well as any number of
+items to be contextually "placed next to" the initial declaration.
+
+Again, though, there is no hygiene, it's as if the source was simply
+copy/pasted. All span information for the `TokenStream` structures returned by
+`from_source` will point to the original `#[derive]` annotation. This means
+that error messages related to struct definitions will get *worse* if they have
+a custom derive attribute placed on them, because the entire struct's span will
+get folded into the `#[derive]` annotation. Eventually, though, more span
+information will be stable on the `TokenStream` type, so this is just a
+temporary limitation.
+
+The `rustc_macro_derive` attribute requires the signature (similar to [macros
+2.0][mac20sig]):
+
+[mac20sig]: http://ncameron.org/blog/libmacro/#tokenisingandquasiquoting
+
+```rust
+fn(&mut Context, TokenStream) -> TokenStream
+```
+
+If a macro cannot process the input token stream, it is expected to panic for
+now, although eventually it will call methods on `Context` to provide more
+structured errors. The compiler will wrap up the panic message and display it
+to the user appropriately. Eventually, however, `librustc_macro` will provide
+more interesting methods of signaling errors to users.
+
+Customization of user-defined `#[derive]` modes can still be done through custom
+attributes, although it will be required for `rustc_macro_derive`
+implementations to remove these attributes when handing them back to the
+compiler. The compiler will still gate unknown attributes by default.
+
+### `rustc-macro` crates
+
+Like the rlib and dylib crate types, the `rustc-macro` crate
+type is intended to be an intermediate product.  What it *actually* produces is
+not specified, but if a `-L` path is provided to it then the compiler will
```

Comment:
This feels a bit of an abuse of "link" if we switch plugins to IPC executables - maybe a separate flag would be a good idea?